File: firmware/rpi/mods/basic_io.js
import { addCallback, piGpio, NO_RETURN, USE_RETURN, ERR_RETURN } from './fw_common.js';

const basic_io_group = "bio";

var pins = {};

export function setup() {
    addCallback(basic_io_group, "", "mn", modelName);
    addCallback(basic_io_group, "", "pm", pinMode);
    addCallback(basic_io_group, "", "dw", digitalWrite);
    addCallback(basic_io_group, "", "dr", digitalRead);
    addCallback(basic_io_group, "", "ar", analogRead);
}

export function shutdown() {
    for (let p_no in pins) {
        pins[p_no] = null;
        delete pins[p_no];
    }
}

async function modelName(param) {
    return { "s": USE_RETURN, "v": "rpi" };
}

async function pinMode(param) {
    const p_no = param.p;
    const mode = param.m;

    try {
        let pin;
        if (p_no in pins) {
            pin = pins[p_no];
        }
        else {
            pin = new piGpio(p_no);
        }
        if (mode == 0) {
            pin.mode(piGpio.OUTPUT);
        }
        else if (mode == 1) {
            pin.mode(piGpio.INPUT);
            pin.pullUpDown(piGpio.PUD_OFF);
        }
        else if (mode == 2) {
            pin.mode(piGpio.INPUT);
            pin.pullUpDown(piGpio.PUD_DOWN);
        }
        pins[p_no] = pin;
        return { "s": NO_RETURN };
    
    }
    catch (e) {
        return { "s": ERR_RETURN, "m": e.toString() };
    }
}

async function digitalWrite(param) {
    const p_no = param.p;
    const value = param.v;

    try {
        let pin = pins[p_no];
        pin.digitalWrite(value);
        return { "s": NO_RETURN };
    }
    catch (e) {
        return { "s": ERR_RETURN, "m": e.toString() };
    }
}

async function digitalRead(param) {
    const p_no = param.p;

    try {
        let pin = pins[p_no];
        let v = pin.digitalRead();
        return { "s": USE_RETURN, "v": v };
    }
    catch (e) {
        return { "s": ERR_RETURN, "m": e.toString() };
    }
}

async function analogRead(param) {
    return { "s": USE_RETURN, "v": 0 };
}

File: firmware/tmkfirm/basic_io.cpp
#include "fw_common.h"
#include "basic_io.h"

#if defined(ESP32)
const char* modelname = "esp32";
#elif defined(ARDUINO_ARCH_RP2040)
const char* modelname = "picow";
#endif

String basic_io_group = "bio";

void basic_io_setup(void) {
  addCallback(basic_io_group, "", "mn", &basic_io_modelname);
  addCallback(basic_io_group, "", "pm", &basic_io_pinMode);
  addCallback(basic_io_group, "", "dw", &basic_io_digitalWrite);
  addCallback(basic_io_group, "", "dr", &basic_io_digitalRead);
  addCallback(basic_io_group, "", "ar", &basic_io_analogRead);
}

void basic_io_modelname(JSONVar &p) {
  r_stat["s"] = (int) USE_RETURN;
  r_stat["t"] = (int) TYPE_STRING;
  r_stat["v"] = modelname;
}

void basic_io_pinMode(JSONVar &p) {
  int p_no, mode;

  p_no = (int) p["p"];
  mode = (int) p["m"];
  switch (mode) {
    case 0:
      pinMode(p_no, OUTPUT);
      break;
    case 1:
      pinMode(p_no, INPUT);
      break;
    case 2:
      pinMode(p_no, INPUT_PULLUP);
      break;
  }
  r_stat["s"] = (int) NO_RETURN;
}

void basic_io_digitalWrite(JSONVar &p) {
  int p_no, value;

  p_no = (int) p["p"];
  value = (int) p["v"];
  digitalWrite(p_no, value);
  r_stat["s"] = (int) NO_RETURN;
}

void basic_io_digitalRead(JSONVar &p) {
  int p_no, value;

  p_no = (int) p["p"];
  value = digitalRead(p_no);
  r_stat["s"] = (int) USE_RETURN;
  r_stat["t"] = (int) TYPE_INT;
  r_stat["v"] = value;
}

void basic_io_analogRead(JSONVar &p) {
  int p_no, value;

  p_no = (int) p["p"];
  value = analogRead(p_no);
  r_stat["s"] = (int) USE_RETURN;
  r_stat["t"] = (int) TYPE_INT;
  r_stat["v"] = value;
}
